BytePair Encoding - Used to train LLMs such as GPT-2, GPT-3 and original model.
Since implementing BPE can be relatively complicated , we will use an existing open source library.

In [1]:
! pip install tiktoken

In [2]:
import importlib
import tiktoken

print("tiktoken version:",importlib.metadata.version("tiktoken"))

tiktoken version: 0.11.0


Once installed, we can instantiate the BPE tokenizer from tiktoken as follows


The usage is similar to tokenizer used in lecture 7 , simpletokenizzer version2

In [5]:
tokenizer = tiktoken.get_encoding("gpt2")

text = (
    "Hello, do you like tea? <|endoftext|> in the sunlit terraces"
    "of someunknownPlace."
)
integers = tokenizer.encode(text,allowed_special={"<|endoftext|>"})
print(integers)

#someunknownPlace also is encoded as we use bytepair encoding.

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 287, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


We can convert tokenid's back to text using decode method.



In [6]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> in the sunlit terracesof someunknownPlace.


Two noteworthy observations based on the token IDs and decoded text.

First, the <|endoftext|> token is assigned a relatively large token ID, namely 50256.
Infact, the BPE tokenizer , which was used to train models such as GPT-2, GPT-3 a, and the original model with <|endoftext|> being assigned the largest token ID.

Second, the BPE tokenizer above encodes and decodes unknown words , such as "someunknownPlace".The BPE tokenizer can handle any unknown word. How does it achieve this without using <|unk|> tokens?


The algorithm underlying BPE breaks down words that aren't in its predefined vocabulary into smaller subword units or even individual characters.This enables it to handle out of vocabulary words.




In [7]:
# Another example of how BPE deals with unknown words.

integers = tokenizer.encode("Akwirw ier")
print(integers)
strings = tokenizer.decode(integers)
print(strings)

[33901, 86, 343, 86, 220, 959]
Akwirw ier
